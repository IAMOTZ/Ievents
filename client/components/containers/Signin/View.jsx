import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { RegularTopNavigation } from '../../common/TopNavigation';
import Footer from '../../common/Footer';
import { LoadingIcon } from '../../common/LoadingAnimation';
import { BigAlert, SmallAlert } from '../../common/Alert';

const View = props => (
  <div id="sign-in-container">
    <RegularTopNavigation />
    <div className="d-flex flex-column align-items-center page-content">
      <LoadingIcon start={props.loggingUserStarted} size={3} />
      <div className="card card-form">
        <h1 className="card-header">Log in</h1>
        <div className="card-body">
          <BigAlert message={props.loggingUserError ? 'Email or password incorrect' : null} />
          <form>
            <div className="form-group">
              <label htmlFor="email">Email</label>
              <div className="input-group">
                <div className="input-group-addon">
                  <i className="fa fa-envelope" />
                </div>
                <input
                  type="email"
                  className="form-control"
                  id="email"
                  placeholder="Your Email"
                  name="email"
                  onChange={props.getInput}
                />
              </div>
              <SmallAlert message={props.inputErrors.emailError} />
            </div>
            <div className="form-group">
              <label htmlFor="password">Password</label>
              <div className="input-group">
                <div className="input-group-addon">
                  <i className="fa fa-user-secret" />
                </div>
                <input
                  type="password"
                  className="form-control"
                  id="password"
                  placeholder="Password"
                  name="password"
                  onChange={props.getInput}
                />
              </div>
              <SmallAlert message={props.inputErrors.passwordError} />
            </div>
            <button
              className="btn btn-block ie-blue-button text-white"
              disabled={props.loggingUserStarted}
              onClick={props.login}
            >Log in
            </button>
            <div className="text-center mt-2" style={{ fontSize: '15px' }}>
              <span>{'Don\'t have an account?'}</span>&nbsp;
              <Link to="/users/signup">Sign up</Link>
            </div>
          </form>
        </div>
      </div>
    </div>
    <Footer />
  </div>
);

View.defaultProps = {
  loggingUserError: '',
  inputErrors: {},
};

/* eslint-disable react/forbid-prop-types */
View.propTypes = {
  loggingUserStarted: PropTypes.bool.isRequired,
  loggingUserError: PropTypes.string,
  inputErrors: PropTypes.object,
  getInput: PropTypes.func.isRequired,
  login: PropTypes.func.isRequired,
};

export default View;
