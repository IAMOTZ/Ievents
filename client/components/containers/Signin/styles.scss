$custom-black: rgba(0,0,0, 0.9);
$custom-blue: rgb(0, 142, 214);

#sign-in-container {
  .page-content {
    margin-top: 100px;
    min-height: 70vh;
  }
  .card-form {
    margin-bottom: 95px;
    width: 350px;
    .card-title {
      background-color: $custom-black;
      color: white;
    }
  }
}