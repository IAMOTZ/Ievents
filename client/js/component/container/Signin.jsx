import React from 'react';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';

import signupStyles from '../../../sass/signup.scss';
import TopNavigation from '../common/TopNavigation.jsx';
import Footer from '../common/Footer.jsx';
import { WarningAlert } from '../common/Alert';

import { loginUser, clearError } from '../../actions/authAction';

@connect((store) => {
  return {
    user: store.user,
    error: store.user.status.error.message,    
  }
})

export default class Signin extends React.Component {
  constructor() {
    super();
    this.state = {
      email: null,
      password: null,
    };
  }

  componentWillUnmount() {
    this.props.dispatch(clearError());
  }

  // This method uses user input to update the state
  getInput = (e) => {
    const state = this.state;
    state[e.target.name] = e.target.value;
    this.setState(state);
  }

  // This method fires the action to log in a user
  login = () => {
    const {
     email, password,
    } = this.state;
    this.props.dispatch(loginUser({
      email,
      password
    }));
  }

  render() {
    if (this.props.user.status.fetched === true) {
      return (
        <div>
          <Redirect to="/addEvent" />
        </div>
      );
    } else {
      return (
        <div>
          <TopNavigation />
          <div className="the-flex-box d-flex flex-column align-items-center">
            <div className="card card-form">
              <WarningAlert message={this.props.error}/>
              <h1 className="card-header">Sign in</h1>
              <div className="card-body">
                <form>
                  <div className="form-group">
                    <label for="email">Email</label>
                    <div className="input-group">
                      <div className="input-group-addon">
                        <i className="fa fa-envelope"></i>
                      </div>
                      <input type="email" className="form-control" id="email" placeholder="Your Email" name="email" onChange={this.getInput} />
                    </div>
                  </div>
                  <div className="form-group">
                    <label for="password">Password</label>
                    <div className="input-group">
                      <div className="input-group-addon">
                        <i className="fa fa-user-secret"></i>
                      </div>
                      <input type="password" className="form-control" id="password" placeholder="Password" name="password" onChange={this.getInput} />
                    </div>
                  </div>
                  <a className="btn btn-block dark-button text-white" onClick={this.login}>Log in</a>
                  <div className="text-center mt-2">
                    <a href="" className="text-muted">forgot password?</a>
                  </div>
                </form>
              </div>
            </div>

            <Footer />
          </div>

        </div>
      );
    }

  }
}
