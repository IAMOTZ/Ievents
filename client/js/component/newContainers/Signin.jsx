import React from 'react';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { stopAsyncProcess } from '../../actions/commonActions';
import * as asyncProcess from '../../actions/asyncProcess';
import { validateSigninInputs } from '../../helpers/inputValidators';
import { loginUser } from '../../actions/authAction';
import TopNavigation from '../common/TopNavigation';
import Footer from '../common/Footer';
import { LoadingIcon } from '../common/LoadingAnimation';
import { BigAlert, SmallAlert } from '../common/Alert';

@connect(store => (
  {
    user: store.authReducer.user,
    loggingUserStarted: store.authReducer.loggingUserStarted,
    loggingUserResolved: store.authReducer.loggingUserResolved,
    loggingUserError: store.authReducer.loggingUserError,
  }
))
class Signin extends React.Component {
  constructor() {
    super();
    this.state = {
      email: null,
      password: null,
      inputErrors: {
        emailError: null,
        passwordError: null,
      },
    };
  }

  componentWillUnmount() {
    this.props.dispatch(stopAsyncProcess(asyncProcess.LOGGING_USER));
  }

  /**
   * Stores the user inputs in the state of this component.
   * @param {Event} event The event object.
   */
  getInput = (event) => {
    const state = { ...this.state };
    state[event.target.name] = event.target.value;
    this.setState(state);
  }

  /**
   * Dispatches an action to authenticate a user.
   */
  login = (event) => {
    event.preventDefault();
    this.props.dispatch(stopAsyncProcess(asyncProcess.LOGGING_USER));
    const { email, password } = this.state;
    const inputErrors = validateSigninInputs(this.state);
    if (inputErrors.errorFound) {
      const state = { ...this.state };
      state.inputErrors = inputErrors;
      this.setState(state);
    } else {
      this.clearInputErrors();
      this.props.dispatch(loginUser({ email, password }));
    }
  }

  clearInputErrors = () => {
    const state = { ...this.state };
    state.inputErrors = {
      emailError: null,
      passwordError: null,
    };
    this.setState(state);
  }

  render() {
    if (this.props.loggingUserResolved) {
      return (<Redirect to="/addEvent" />);
    }
    return (
      <div id="sign-in-container">
        <TopNavigation />
        <div className="d-flex flex-column align-items-center main-content">
          <LoadingIcon start={this.props.loggingUserStarted} size={3} />
          <div className="card card-form">
            <h1 className="card-header">Sign in</h1>
            <div className="card-body">
              <BigAlert message={this.props.loggingUserError ? 'email or password incorrect' : null} />
              <form>
                <div className="form-group">
                  <label htmlFor="email">Email</label>
                  <div className="input-group">
                    <div className="input-group-addon">
                      <i className="fa fa-envelope" />
                    </div>
                    <input
                      type="email"
                      className="form-control"
                      id="email"
                      placeholder="Your Email"
                      name="email"
                      onChange={this.getInput}
                    />
                  </div>
                  <SmallAlert message={this.state.inputErrors.emailError} />
                </div>
                <div className="form-group">
                  <label htmlFor="password">Password</label>
                  <div className="input-group">
                    <div className="input-group-addon">
                      <i className="fa fa-user-secret" />
                    </div>
                    <input
                      type="password"
                      className="form-control"
                      id="password"
                      placeholder="Password"
                      name="password"
                      onChange={this.getInput}
                    />
                  </div>
                  <SmallAlert message={this.state.inputErrors.passwordError} />
                </div>
                <button
                  className="btn btn-block dark-button text-white"
                  disabled={this.props.loggingUserStarted}
                  onClick={this.login}
                >Log in
                  </button>
                <div className="text-center mt-2">
                  <a href="" className="text-muted">forgot password?</a>
                </div>
              </form>
            </div>
          </div>
          <Footer />
        </div>
      </div>
    );
  }
}

export default Signin;
