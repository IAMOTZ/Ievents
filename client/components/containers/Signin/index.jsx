import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { validateSigninInputs } from '../../../helpers/inputValidators';
import { loginUser } from '../../../actions/authAction';
import { stopAsyncProcess } from '../../../actions/commonActions';
import * as asyncProcess from '../../../actions/asyncProcess';
import './styles.scss';
import View from './View';

@connect(store => (
  {
    loggingUserStarted: store.authReducer.loggingUserStarted,
    loggingUserResolved: store.authReducer.loggingUserResolved,
    loggingUserError: store.authReducer.loggingUserError,
  }
))
class Signin extends React.Component {
  constructor() {
    super();
    this.state = {
      email: null,
      password: null,
      inputErrors: {
        emailError: null,
        passwordError: null,
      },
    };
  }

  componentWillUnmount() {
    this.props.dispatch(stopAsyncProcess(asyncProcess.LOGGING_USER));
  }

  /**
   * Stores the user inputs in the state of this component.
   * @param {Event} event The event object.
   */
  getInput = (event) => {
    const state = { ...this.state };
    state[event.target.name] = event.target.value;
    this.setState(state);
  }

  /**
   * Dispatches an action to authenticate a user.
   */
  login = (event) => {
    event.preventDefault();
    this.props.dispatch(stopAsyncProcess(asyncProcess.LOGGING_USER));
    const { email, password } = this.state;
    const inputErrors = validateSigninInputs(this.state);
    if (inputErrors.errorFound) {
      const state = { ...this.state };
      state.inputErrors = inputErrors;
      this.setState(state);
    } else {
      this.clearInputErrors();
      this.props.dispatch(loginUser({ email, password }));
    }
  }

  clearInputErrors = () => {
    const state = { ...this.state };
    state.inputErrors = {
      emailError: null,
      passwordError: null,
    };
    this.setState(state);
  }

  render() {
    if (this.props.loggingUserResolved) {
      return (<Redirect to="/events" />);
    }
    return (
      <View
        loggingUserStarted={this.props.loggingUserStarted}
        loggingUserError={this.props.loggingUserError}
        inputErrors={this.state.inputErrors}
        getInput={this.getInput}
        login={this.login}
      />);
  }
}


Signin.defaultProps = {
  loggingUserStarted: false,
  loggingUserResolved: false,
  loggingUserError: '',
  dispatch: () => {},
};

Signin.propTypes = {
  loggingUserStarted: PropTypes.bool,
  loggingUserResolved: PropTypes.bool,
  loggingUserError: PropTypes.string,
  dispatch: PropTypes.func,
};

export default Signin;
