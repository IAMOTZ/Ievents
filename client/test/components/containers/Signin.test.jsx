import React from 'react';
import Signin from '../../../components/containers/Signin';
import SigninView from '../../../components/containers/Signin/View';

const mockEvent = target => ({
  preventDefault: jest.fn(), target,
});

const sampleUserData = {
  email: 'tester@gmail.com',
  password: 'Password123',
};

describe('<Signin />', () => {
  const props = {
    dispatch: jest.fn(),
    loggingUserStarted: false,
    loggingUserResolved: false,
    loggingUserError: null,
  };
  const alterProps = newProps => ({ ...props, ...newProps });
  describe('Rendering:', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<Signin {...props} />);
    });
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should redirect to centers page if logging resolved', () => {
      wrapper.setProps(alterProps({ loggingUserResolved: true }));
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Behaviour:', () => {
    let wrapper; let wrapperInstance;
    beforeEach(() => {
      props.dispatch.mockReset();
      wrapper = shallow(<Signin {...props} />);
      wrapperInstance = wrapper.instance();
    });
    // The names on the following test block refers
    // to the instance method that handles the behaviour tested.
    describe('GET_INPUT', () => {
      it('should update the state with user input', () => {
        const userInput = sampleUserData.email;
        wrapperInstance.getInput(mockEvent({ name: 'email', value: userInput }));
        expect(wrapper.state('email')).toEqual(userInput);
      });
    });
    describe('CLEAR_INPUT_ERRORS', () => {
      it('should clear input errors in the state', () => {
        const emailError = 'You name has errors';
        wrapper.setState({ inputErrors: { emailError } });
        wrapperInstance.clearInputErrors();
        expect(wrapper.state('inputErrors').emailError).toBeNull();
      });
    });

    describe('LOGIN', () => {
      it('should dispatch an action and set inputErrors if validation fails', () => {
        const wrongEmail = 'tester.com';
        wrapper.setState({ email: wrongEmail });
        wrapperInstance.login(mockEvent());
        expect(props.dispatch.mock.calls.length).toEqual(1);
        expect(typeof wrapper.state('inputErrors').emailError).toEqual('string');
      });
      it('should clear inputErrors and dispatch 2 actions if validation pass', () => {
        const previousError = 'Email is required';
        wrapper.setState({
          ...sampleUserData, inputErrors: { emailError: previousError }
        });
        wrapperInstance.login(mockEvent());
        expect(wrapper.state('inputErrors').emailError).toBeNull();
        expect(props.dispatch.mock.calls.length).toEqual(2);
      });
    });

    describe('COMOPONENT_WILL_UNMOUNT', () => {
      it('should dispatch an action when the component wants to unmount', () => {
        wrapper.unmount();
        expect(props.dispatch.mock.calls.length).toEqual(1);
      });
    });
  });
});

describe('<SigninView />', () => {
  const props = {
    getInput: jest.fn(),
    inputErrors: {},
    loggingUserError: null,
    loggingUserStarted: false,
    login: jest.fn(),
  };
  describe('Rendering', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<SigninView {...props} />);
    });
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
