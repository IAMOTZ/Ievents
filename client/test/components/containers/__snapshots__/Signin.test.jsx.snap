// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signin /> Rendering: should redirect to centers page if logging resolved 1`] = `
<Redirect
  push={false}
  to="/centers"
/>
`;

exports[`<Signin /> Rendering: should render correctly 1`] = `
<View
  getInput={[Function]}
  inputErrors={
    Object {
      "emailError": null,
      "passwordError": null,
    }
  }
  loggingUserError={null}
  loggingUserStarted={false}
  login={[Function]}
/>
`;

exports[`<SigninView /> Rendering should render correctly 1`] = `
<div
  id="sign-in-container"
>
  <RegularTopNavigation />
  <div
    className="d-flex flex-column align-items-center page-content"
  >
    <LoadingIcon
      size={3}
      start={false}
    />
    <div
      className="card card-form"
    >
      <h1
        className="card-header"
      >
        Log in
      </h1>
      <div
        className="card-body"
      >
        <BigAlert
          message={null}
        />
        <form>
          <div
            className="form-group"
          >
            <label
              htmlFor="email"
            >
              Email
            </label>
            <div
              className="input-group"
            >
              <div
                className="input-group-addon"
              >
                <i
                  className="fa fa-envelope"
                />
              </div>
              <input
                className="form-control"
                id="email"
                name="email"
                onChange={[MockFunction]}
                placeholder="Your Email"
                type="email"
              />
            </div>
            <SmallAlert />
          </div>
          <div
            className="form-group"
          >
            <label
              htmlFor="password"
            >
              Password
            </label>
            <div
              className="input-group"
            >
              <div
                className="input-group-addon"
              >
                <i
                  className="fa fa-user-secret"
                />
              </div>
              <input
                className="form-control"
                id="password"
                name="password"
                onChange={[MockFunction]}
                placeholder="Password"
                type="password"
              />
            </div>
            <SmallAlert />
          </div>
          <button
            className="btn btn-block ie-blue-button text-white"
            disabled={false}
            onClick={[MockFunction]}
          >
            Log in
          </button>
          <div
            className="text-center mt-2"
            style={
              Object {
                "fontSize": "15px",
              }
            }
          >
            <span>
              Don't have an account?
            </span>
            Â 
            <Link
              replace={false}
              to="/users/signup"
            >
              Sign up
            </Link>
          </div>
        </form>
      </div>
    </div>
  </div>
  <Footer />
</div>
`;
